### ISSUE TITLE
lineinfile: string is converted to python dictionary

### ISSUE THREAD

**Issue by photoninger** [ansible/ansible#21065]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
lineinfile

##### ANSIBLE VERSION
```
ansible 2.2.1.0
```
and
```
ansible 2.3.0 (devel db8fd95d68) last updated 2017/02/06 16:21:29 (GMT +200)
```
##### CONFIGURATION
nothing special

##### OS / ENVIRONMENT
Running on macOS 10.12 against SLES-12.2

##### SUMMARY
My playbook adds line(s) to a config file. 
The line(s) looks like a python or json dictionary.

##### STEPS TO REPRODUCE
dest file on server:
```
instances = [
{"port" : 12184, "instance" : "zookeeper-2184"},
{"port" : 12183, "instance" : "zookeeper-2183"},
{"port" : 12182, "instance" : "zookeeper-2182"},
{"port" : 12181, "instance" : "zookeeper-2181"},
]
```

Playbook to show the bug
```yaml
---
- hosts: testserver
  remote_user: root
  vars:
    - ZOOKEEPERPORT: 2222
    - bla: abc
  tasks:
  
  - name: add line to config - integer variable
    lineinfile: 
      dest: /root/test_jolokia.cfg
      line: "{\"port\" : 1{{ ZOOKEEPERPORT }}, \"instance\" : \"zookeeper-{{ ZOOKEEPERPORT }}\"}," 
      state: present 
      insertafter: "^instances = "

  - name: add line to config - string variable
    lineinfile: 
      dest: /root/test_jolokia.cfg
      line: "{\"port\" : 1{{ bla }}, \"instance\" : \"zookeeper-{{ bla }}\"}," 
      state: present 
      insertafter: "^instances = "

  - name: add line to config - fixed string
    lineinfile: 
      dest: /root/test_jolokia.cfg
      line: "{\"port\" : 11111, \"instance\" : \"zookeeper-1111\"}," 
      state: present 
      insertafter: "^instances = "
```

##### EXPECTED RESULTS
There should be 3 additional lines like:
```
instances = [
{"port" : 11111, "instance" : "zookeeper-1111"},
{"port" : 1abc, "instance" : "zookeeper-abc"},
{"port" : 12222, "instance" : "zookeeper-2222"},
{"port" : 12184, "instance" : "zookeeper-2184"},
{"port" : 12183, "instance" : "zookeeper-2183"},
{"port" : 12182, "instance" : "zookeeper-2182"},
{"port" : 12181, "instance" : "zookeeper-2181"},
]
```

##### ACTUAL RESULTS
The line with the integer variable is converted to a list with one dictionary item and then added to the file:
```
instances = [
{"port" : 11111, "instance" : "zookeeper-1111"},
{"port" : 1abc, "instance" : "zookeeper-abc"},
[{'instance': 'zookeeper-2222', 'port': 12222}]
{"port" : 12184, "instance" : "zookeeper-2184"},
{"port" : 12183, "instance" : "zookeeper-2183"},
{"port" : 12182, "instance" : "zookeeper-2182"},
{"port" : 12181, "instance" : "zookeeper-2181"},
]
```

In the output of ansible-playbook -vvv the three lines appear as:
```
[...]
"line": "[{'instance': 'zookeeper-2222', 'port': 12222}]", 
[...]
"line": "{\"port\" : 1abc, \"instance\" : \"zookeeper-abc\"},", 
[...]
"line": "{\"port\" : 11111, \"instance\" : \"zookeeper-1111\"},", 
[...]
```
##### WORKAROUND
My workaround is to add one space to the beginning of the line, then it works as it should:
```yaml
---
[...]
  - name: add line to config - integer variable
    lineinfile: 
      dest: /root/test_jolokia.cfg
      line: " {\"port\" : 1{{ ZOOKEEPERPORT }}, \"instance\" : \"zookeeper-{{ ZOOKEEPERPORT }}\"}," 
      state: present 
      insertafter: "^instances = "
[...]
```
dest file:
```
instances = [
{"port" : 11111, "instance" : "zookeeper-1111"},
{"port" : 1abc, "instance" : "zookeeper-abc"},
 {"port" : 12222, "instance" : "zookeeper-2222"},
{"port" : 12184, "instance" : "zookeeper-2184"},
{"port" : 12183, "instance" : "zookeeper-2183"},
{"port" : 12182, "instance" : "zookeeper-2182"},
{"port" : 12181, "instance" : "zookeeper-2181"},
]
```
The fixed line in verbose output:
```
"line": " {\"port\" : 12222, \"instance\" : \"zookeeper-2222\"},",
```

**Comment by alikins** [comment]
I can reproduce this with devel.

A version of example playbook with added debugging:

``` yaml
--
- hosts: localhost
  gather_facts: false
  vars:
    ZOOKEEPERPORT: 2222
    bla: abc
    an_int: 12
    line_fails: "{\"port\" : 1{{ ZOOKEEPERPORT }}, \"instancef\" : \"zookeeper-{{ ZOOKEEPERPORT }}\"},"
    line_works: "{\"port\" : 1{{ bla }}, \"instance\" : \"zookeeper-{{ bla }}\"},"
    simple_line: "{\"a\" : {{ an_int }}, \"b\" : \"c-{{ an_int }}\"},"
  tasks:

  - name: create test dir
    file:
        dest: /tmp/test_21065
        state: directory

  - name: cp test file
    copy:
        src: test_lineinfile.cfg
        dest: /tmp/test_21065/test_lineinfile.cfg

  - name: debug vars
    debug:
      var: ZOOKEEPERPORT

  - name: debug vars msg
    debug:
      msg: "{{ ZOOKEEPERPORT }} type {{ ZOOKEEPERPORT|type_debug }}"

  - name: debug line_fails
    debug:
      msg: "{{ line_fails }} type {{ line_fails|type_debug }}"

  - name: debug line_works
    debug:
      msg: "{{ line_works }} type {{ line_works|type_debug }}"

  - name: debug simple_line
    debug:
      msg: "{{ simple_line }} type {{ simple_line|type_debug }}"

  - name: add line to config - integer variable
    lineinfile:
      dest: /tmp/test_21065/test_lineinfile.cfg
      line: "{\"port\" : 1{{ ZOOKEEPERPORT }}, \"instance\" : \"zookeeper-{{ ZOOKEEPERPORT }}\"},"
      state: present
      insertafter: "^instances = "

  - name: add line to config - string variable
    lineinfile:
      dest: /tmp/test_21065/test_lineinfile.cfg
      line: "{\"port\" : 1{{ bla }}, \"instance\" : \"zookeeper-{{ bla }}\"},"
      state: present
      insertafter: "^instances = "

  - name: add line to config - fixed string
    lineinfile:
      dest: /tmp/test_21065/test_lineinfile.cfg
      line: "{\"port\" : 11111, \"instance\" : \"zookeeper-1111\"},"
      state: present
      insertafter: "^instances = "
```

output:

``` 
[fedora-25:lineinfile_json_21065 (master %)]$ ANSIBLE_KEEP_REMOTE_FILES=1 ansible-playbook --diff -vvv test.yml
Using /home/adrian/ansible/ansible.cfg as config file
 [WARNING]: provided hosts list is empty, only localhost is available


PLAYBOOK: test.yml *******************************************************************************************************************************************************************************************************************************************
1 plays in test.yml

PLAY [localhost] *********************************************************************************************************************************************************************************************************************************************
META: ran handlers

TASK [create test dir] ***************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible-bug-repro/lineinfile_json_21065/test.yml:13
Using module file /home/adrian/src/ansible/lib/ansible/modules/files/file.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1486403121.68-23247120886173 `" && echo ansible-tmp-1486403121.68-23247120886173="` echo /home/adrian/.ansible/tmp/ansible-tmp-1486403121.68-23247120886173 `" ) && sleep 0'                                                                                                                                                                                                                                              
<127.0.0.1> PUT /tmp/tmpUFYD_X TO /home/adrian/.ansible/tmp/ansible-tmp-1486403121.68-23247120886173/file.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1486403121.68-23247120886173/ /home/adrian/.ansible/tmp/ansible-tmp-1486403121.68-23247120886173/file.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1486403121.68-23247120886173/file.py && sleep 0'
ok: [localhost] => {
    "_ansible_no_log": false, 
    "_ansible_parsed": true, 
    "changed": false, 
    "diff": {
        "after": {
            "path": "/tmp/test_21065"
        }, 
        "before": {
            "path": "/tmp/test_21065"
        }
    }, 
    "gid": 1000, 
    "group": "adrian", 
    "invocation": {
        "module_args": {
            "attributes": null, 
            "backup": null, 
            "content": null, 
            "delimiter": null, 
            "dest": "/tmp/test_21065", 
            "diff_peek": null, 
            "directory_mode": null, 
            "follow": false, 
            "force": false, 
            "group": null, 
            "mode": null, 
            "original_basename": null, 
            "owner": null, 
            "path": "/tmp/test_21065", 
            "recurse": false, 
            "regexp": null, 
            "remote_src": null, 
            "selevel": null, 
            "serole": null, 
            "setype": null, 
            "seuser": null, 
            "src": null, 
            "state": "directory", 
            "unsafe_writes": null, 
            "validate": null
        }, 
        "module_name": "file"
    }, 
    "mode": "0775", 
    "owner": "adrian", 
    "path": "/tmp/test_21065", 
    "secontext": "unconfined_u:object_r:user_tmp_t:s0", 
    "size": 60, 
    "state": "directory", 
    "uid": 1000
}

TASK [cp test file] ******************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible-bug-repro/lineinfile_json_21065/test.yml:18
Using module file /home/adrian/src/ansible/lib/ansible/modules/files/stat.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1486403121.87-36735013705944 `" && echo ansible-tmp-1486403121.87-36735013705944="` echo /home/adrian/.ansible/tmp/ansible-tmp-1486403121.87-36735013705944 `" ) && sleep 0'                                                                                                                                                                                                                                              
<127.0.0.1> PUT /tmp/tmpis00Fr TO /home/adrian/.ansible/tmp/ansible-tmp-1486403121.87-36735013705944/stat.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1486403121.87-36735013705944/ /home/adrian/.ansible/tmp/ansible-tmp-1486403121.87-36735013705944/stat.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1486403121.87-36735013705944/stat.py && sleep 0'
Using module file /home/adrian/src/ansible/lib/ansible/modules/files/file.py
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1486403121.99-124024532842003 `" && echo ansible-tmp-1486403121.99-124024532842003="` echo /home/adrian/.ansible/tmp/ansible-tmp-1486403121.99-124024532842003 `" ) && sleep 0'                                                                                                                                                                                                                                           
<127.0.0.1> PUT /tmp/tmpTZNYBH TO /home/adrian/.ansible/tmp/ansible-tmp-1486403121.99-124024532842003/file.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1486403121.99-124024532842003/ /home/adrian/.ansible/tmp/ansible-tmp-1486403121.99-124024532842003/file.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1486403121.99-124024532842003/file.py && sleep 0'
ok: [localhost] => {
    "_ansible_no_log": false, 
    "_ansible_parsed": true, 
    "changed": false, 
    "checksum": "473c46f801be3863e3b34abe66cf394d1a53c0d7", 
    "dest": "/tmp/test_21065/test_lineinfile.cfg", 
    "diff": {
        "after": {
            "path": "/tmp/test_21065/test_lineinfile.cfg"
        }, 
        "before": {
            "path": "/tmp/test_21065/test_lineinfile.cfg"
        }
    }, 
    "gid": 1000, 
    "group": "adrian", 
    "invocation": {
        "module_args": {
            "attributes": null, 
            "backup": null, 
            "content": null, 
            "delimiter": null, 
            "dest": "/tmp/test_21065/test_lineinfile.cfg", 
            "diff_peek": null, 
            "directory_mode": null, 
            "follow": false, 
            "force": false, 
            "group": null, 
            "mode": null, 
            "original_basename": "test_lineinfile.cfg", 
            "owner": null, 
            "path": "/tmp/test_21065/test_lineinfile.cfg", 
            "recurse": false, 
            "regexp": null, 
            "remote_src": null, 
            "selevel": null, 
            "serole": null, 
            "setype": null, 
            "seuser": null, 
            "src": "test_lineinfile.cfg", 
            "state": null, 
            "unsafe_writes": null, 
            "validate": null
        }
    }, 
    "mode": "0664", 
    "owner": "adrian", 
    "path": "/tmp/test_21065/test_lineinfile.cfg", 
    "secontext": "unconfined_u:object_r:user_tmp_t:s0", 
    "size": 212, 
    "state": "file", 
    "uid": 1000
}

TASK [debug vars] ********************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible-bug-repro/lineinfile_json_21065/test.yml:23
ok: [localhost] => {
    "ZOOKEEPERPORT": 2222, 
    "_ansible_no_log": false, 
    "_ansible_verbose_always": true, 
    "changed": false, 
    "invocation": {
        "module_args": {
            "var": "ZOOKEEPERPORT"
        }, 
        "module_name": "debug"
    }
}

TASK [debug vars msg] ****************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible-bug-repro/lineinfile_json_21065/test.yml:27
ok: [localhost] => {
    "_ansible_no_log": false, 
    "_ansible_verbose_always": true, 
    "changed": false, 
    "invocation": {
        "module_args": {
            "msg": "2222 type int"
        }, 
        "module_name": "debug"
    }, 
    "msg": "2222 type int"
}

TASK [debug line_fails] **************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible-bug-repro/lineinfile_json_21065/test.yml:31
ok: [localhost] => {
    "_ansible_no_log": false, 
    "_ansible_verbose_always": true, 
    "changed": false, 
    "invocation": {
        "module_args": {
            "msg": "({'instancef': 'zookeeper-2222', 'port': 12222},) type tuple"
        }, 
        "module_name": "debug"
    }, 
    "msg": "({'instancef': 'zookeeper-2222', 'port': 12222},) type tuple"
}

TASK [debug line_works] **************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible-bug-repro/lineinfile_json_21065/test.yml:35
ok: [localhost] => {
    "_ansible_no_log": false, 
    "_ansible_verbose_always": true, 
    "changed": false, 
    "invocation": {
        "module_args": {
            "msg": "{\"port\" : 1abc, \"instance\" : \"zookeeper-abc\"}, type unicode"
        }, 
        "module_name": "debug"
    }, 
    "msg": "{\"port\" : 1abc, \"instance\" : \"zookeeper-abc\"}, type unicode"
}

TASK [debug simple_line] *************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible-bug-repro/lineinfile_json_21065/test.yml:39
ok: [localhost] => {
    "_ansible_no_log": false, 
    "_ansible_verbose_always": true, 
    "changed": false, 
    "invocation": {
        "module_args": {
            "msg": "({'a': 12, 'b': 'c-12'},) type tuple"
        }, 
        "module_name": "debug"
    }, 
    "msg": "({'a': 12, 'b': 'c-12'},) type tuple"
}

TASK [add line to config - integer variable] *****************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible-bug-repro/lineinfile_json_21065/test.yml:43
Using module file /home/adrian/src/ansible/lib/ansible/modules/files/lineinfile.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1486403122.34-195662249850479 `" && echo ansible-tmp-1486403122.34-195662249850479="` echo /home/adrian/.ansible/tmp/ansible-tmp-1486403122.34-195662249850479 `" ) && sleep 0'                                                                                                                                                                                                                                           
<127.0.0.1> PUT /tmp/tmpE0Tmnf TO /home/adrian/.ansible/tmp/ansible-tmp-1486403122.34-195662249850479/lineinfile.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1486403122.34-195662249850479/ /home/adrian/.ansible/tmp/ansible-tmp-1486403122.34-195662249850479/lineinfile.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1486403122.34-195662249850479/lineinfile.py && sleep 0'
--- before: /tmp/test_21065/test_lineinfile.cfg (content)
+++ after: /tmp/test_21065/test_lineinfile.cfg (content)
@@ -1,4 +1,5 @@
 instances = [
+[{'instance': 'zookeeper-2222', 'port': 12222}]
 {"port" : 12184, "instance" : "zookeeper-2184"},
 {"port" : 12183, "instance" : "zookeeper-2183"},
 {"port" : 12182, "instance" : "zookeeper-2182"},

changed: [localhost] => {
    "_ansible_no_log": false, 
    "_ansible_parsed": true, 
    "backup": "", 
    "changed": true, 
    "diff": [
        {
            "after": "instances = [\n[{'instance': 'zookeeper-2222', 'port': 12222}]\n{\"port\" : 12184, \"instance\" : \"zookeeper-2184\"},\n{\"port\" : 12183, \"instance\" : \"zookeeper-2183\"},\n{\"port\" : 12182, \"instance\" : \"zookeeper-2182\"},\n{\"port\" : 12181, \"instance\" : \"zookeeper-2181\"},\n]\n",                                                                                                                                                                                                 
            "after_header": "/tmp/test_21065/test_lineinfile.cfg (content)", 
            "before": "instances = [\n{\"port\" : 12184, \"instance\" : \"zookeeper-2184\"},\n{\"port\" : 12183, \"instance\" : \"zookeeper-2183\"},\n{\"port\" : 12182, \"instance\" : \"zookeeper-2182\"},\n{\"port\" : 12181, \"instance\" : \"zookeeper-2181\"},\n]\n",                                                                                                                                                                                                                                                 
            "before_header": "/tmp/test_21065/test_lineinfile.cfg (content)"
        }, 
        {
            "after_header": "/tmp/test_21065/test_lineinfile.cfg (file attributes)", 
            "before_header": "/tmp/test_21065/test_lineinfile.cfg (file attributes)"
        }
    ], 
    "invocation": {
        "module_args": {
            "attributes": null, 
            "backrefs": false, 
            "backup": false, 
            "content": null, 
            "create": false, 
            "delimiter": null, 
            "dest": "/tmp/test_21065/test_lineinfile.cfg", 
            "directory_mode": null, 
            "follow": false, 
            "force": null, 
            "group": null, 
            "insertafter": "^instances = ", 
            "insertbefore": null, 
            "line": "[{'instance': 'zookeeper-2222', 'port': 12222}]", 
            "mode": null, 
            "owner": null, 
            "path": "/tmp/test_21065/test_lineinfile.cfg", 
            "regexp": null, 
            "remote_src": null, 
            "selevel": null, 
            "serole": null, 
            "setype": null, 
            "seuser": null, 
            "src": null, 
            "state": "present", 
            "unsafe_writes": null, 
            "validate": null
        }, 
        "module_name": "lineinfile"
    }, 
    "msg": "line added"
}

TASK [add line to config - string variable] ******************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible-bug-repro/lineinfile_json_21065/test.yml:50
Using module file /home/adrian/src/ansible/lib/ansible/modules/files/lineinfile.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1486403122.48-201478990463298 `" && echo ansible-tmp-1486403122.48-201478990463298="` echo /home/adrian/.ansible/tmp/ansible-tmp-1486403122.48-201478990463298 `" ) && sleep 0'                                                                                                                                                                                                                                           
<127.0.0.1> PUT /tmp/tmpHlTlgf TO /home/adrian/.ansible/tmp/ansible-tmp-1486403122.48-201478990463298/lineinfile.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1486403122.48-201478990463298/ /home/adrian/.ansible/tmp/ansible-tmp-1486403122.48-201478990463298/lineinfile.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1486403122.48-201478990463298/lineinfile.py && sleep 0'
--- before: /tmp/test_21065/test_lineinfile.cfg (content)
+++ after: /tmp/test_21065/test_lineinfile.cfg (content)
@@ -1,4 +1,5 @@
 instances = [
+{"port" : 1abc, "instance" : "zookeeper-abc"},
 [{'instance': 'zookeeper-2222', 'port': 12222}]
 {"port" : 12184, "instance" : "zookeeper-2184"},
 {"port" : 12183, "instance" : "zookeeper-2183"},

changed: [localhost] => {
    "_ansible_no_log": false, 
    "_ansible_parsed": true, 
    "backup": "", 
    "changed": true, 
    "diff": [
        {
            "after": "instances = [\n{\"port\" : 1abc, \"instance\" : \"zookeeper-abc\"},\n[{'instance': 'zookeeper-2222', 'port': 12222}]\n{\"port\" : 12184, \"instance\" : \"zookeeper-2184\"},\n{\"port\" : 12183, \"instance\" : \"zookeeper-2183\"},\n{\"port\" : 12182, \"instance\" : \"zookeeper-2182\"},\n{\"port\" : 12181, \"instance\" : \"zookeeper-2181\"},\n]\n",                                                                                                                                           
            "after_header": "/tmp/test_21065/test_lineinfile.cfg (content)", 
            "before": "instances = [\n[{'instance': 'zookeeper-2222', 'port': 12222}]\n{\"port\" : 12184, \"instance\" : \"zookeeper-2184\"},\n{\"port\" : 12183, \"instance\" : \"zookeeper-2183\"},\n{\"port\" : 12182, \"instance\" : \"zookeeper-2182\"},\n{\"port\" : 12181, \"instance\" : \"zookeeper-2181\"},\n]\n",                                                                                                                                                                                                
            "before_header": "/tmp/test_21065/test_lineinfile.cfg (content)"
        }, 
        {
            "after_header": "/tmp/test_21065/test_lineinfile.cfg (file attributes)", 
            "before_header": "/tmp/test_21065/test_lineinfile.cfg (file attributes)"
        }
    ], 
    "invocation": {
        "module_args": {
            "attributes": null, 
            "backrefs": false, 
            "backup": false, 
            "content": null, 
            "create": false, 
            "delimiter": null, 
            "dest": "/tmp/test_21065/test_lineinfile.cfg", 
            "directory_mode": null, 
            "follow": false, 
            "force": null, 
            "group": null, 
            "insertafter": "^instances = ", 
            "insertbefore": null, 
            "line": "{\"port\" : 1abc, \"instance\" : \"zookeeper-abc\"},", 
            "mode": null, 
            "owner": null, 
            "path": "/tmp/test_21065/test_lineinfile.cfg", 
            "regexp": null, 
            "remote_src": null, 
            "selevel": null, 
            "serole": null, 
            "setype": null, 
            "seuser": null, 
            "src": null, 
            "state": "present", 
            "unsafe_writes": null, 
            "validate": null
        }, 
        "module_name": "lineinfile"
    }, 
    "msg": "line added"
}

TASK [add line to config - fixed string] *********************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible-bug-repro/lineinfile_json_21065/test.yml:57
Using module file /home/adrian/src/ansible/lib/ansible/modules/files/lineinfile.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1486403122.61-32240335307116 `" && echo ansible-tmp-1486403122.61-32240335307116="` echo /home/adrian/.ansible/tmp/ansible-tmp-1486403122.61-32240335307116 `" ) && sleep 0'                                                                                                                                                                                                                                              
<127.0.0.1> PUT /tmp/tmpRyVnIf TO /home/adrian/.ansible/tmp/ansible-tmp-1486403122.61-32240335307116/lineinfile.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1486403122.61-32240335307116/ /home/adrian/.ansible/tmp/ansible-tmp-1486403122.61-32240335307116/lineinfile.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1486403122.61-32240335307116/lineinfile.py && sleep 0'
--- before: /tmp/test_21065/test_lineinfile.cfg (content)
+++ after: /tmp/test_21065/test_lineinfile.cfg (content)
@@ -1,4 +1,5 @@
 instances = [
+{"port" : 11111, "instance" : "zookeeper-1111"},
 {"port" : 1abc, "instance" : "zookeeper-abc"},
 [{'instance': 'zookeeper-2222', 'port': 12222}]
 {"port" : 12184, "instance" : "zookeeper-2184"},

changed: [localhost] => {
    "_ansible_no_log": false, 
    "_ansible_parsed": true, 
    "backup": "", 
    "changed": true, 
    "diff": [
        {
            "after": "instances = [\n{\"port\" : 11111, \"instance\" : \"zookeeper-1111\"},\n{\"port\" : 1abc, \"instance\" : \"zookeeper-abc\"},\n[{'instance': 'zookeeper-2222', 'port': 12222}]\n{\"port\" : 12184, \"instance\" : \"zookeeper-2184\"},\n{\"port\" : 12183, \"instance\" : \"zookeeper-2183\"},\n{\"port\" : 12182, \"instance\" : \"zookeeper-2182\"},\n{\"port\" : 12181, \"instance\" : \"zookeeper-2181\"},\n]\n", 
            "after_header": "/tmp/test_21065/test_lineinfile.cfg (content)", 
            "before": "instances = [\n{\"port\" : 1abc, \"instance\" : \"zookeeper-abc\"},\n[{'instance': 'zookeeper-2222', 'port': 12222}]\n{\"port\" : 12184, \"instance\" : \"zookeeper-2184\"},\n{\"port\" : 12183, \"instance\" : \"zookeeper-2183\"},\n{\"port\" : 12182, \"instance\" : \"zookeeper-2182\"},\n{\"port\" : 12181, \"instance\" : \"zookeeper-2181\"},\n]\n", 
            "before_header": "/tmp/test_21065/test_lineinfile.cfg (content)"
        }, 
        {
            "after_header": "/tmp/test_21065/test_lineinfile.cfg (file attributes)", 
            "before_header": "/tmp/test_21065/test_lineinfile.cfg (file attributes)"
        }
    ], 
    "invocation": {
        "module_args": {
            "attributes": null, 
            "backrefs": false, 
            "backup": false, 
            "content": null, 
            "create": false, 
            "delimiter": null, 
            "dest": "/tmp/test_21065/test_lineinfile.cfg", 
            "directory_mode": null, 
            "follow": false, 
            "force": null, 
            "group": null, 
            "insertafter": "^instances = ", 
            "insertbefore": null, 
            "line": "{\"port\" : 11111, \"instance\" : \"zookeeper-1111\"},", 
            "mode": null, 
            "owner": null, 
            "path": "/tmp/test_21065/test_lineinfile.cfg", 
            "regexp": null, 
            "remote_src": null, 
            "selevel": null, 
            "serole": null, 
            "setype": null, 
            "seuser": null, 
            "src": null, 
            "state": "present", 
            "unsafe_writes": null, 
            "validate": null
        }, 
        "module_name": "lineinfile"
    }, 
    "msg": "line added"
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=10   changed=3    unreachable=0    failed=0   
```

**Comment by alikins** [comment]
And I don't think this is a bug in lineinfile, it seems to be a general ansible issue.

**Comment by alikins** [comment]
A simpler reproducer not involving lineinfile:

```
[fedora-25:lineinfile_json_21065 (master %)]$ cat test.yml 
---
- hosts: localhost
  gather_facts: false
  vars:
    an_int: 12
    simple_line: "{\"a\" : {{ an_int }}, \"b\" : \"c-{{ an_int }}\"},"
  tasks:

  - name: debug simple_line
    debug:
      msg: "{{ simple_line }} type_debug={{ simple_line|type_debug }}"


  - name: assert that simple_line is string
    assert:
        that: simple_line|type_debug != "tuple"

[fedora-25:lineinfile_json_21065 (master %)]$ ANSIBLE_KEEP_REMOTE_FILES=1 ansible-playbook --diff -vvv test.yml
Using /home/adrian/ansible/ansible.cfg as config file
 [WARNING]: provided hosts list is empty, only localhost is available


PLAYBOOK: test.yml *******************************************************************************************************************************************************************************************************************************************
1 plays in test.yml

PLAY [localhost] *********************************************************************************************************************************************************************************************************************************************
META: ran handlers

TASK [debug simple_line] *************************************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible-bug-repro/lineinfile_json_21065/test.yml:9
ok: [localhost] => {
    "_ansible_no_log": false, 
    "_ansible_verbose_always": true, 
    "changed": false, 
    "invocation": {
        "module_args": {
            "msg": "({'a': 12, 'b': 'c-12'},) type_debug=tuple"
        }, 
        "module_name": "debug"
    }, 
    "msg": "({'a': 12, 'b': 'c-12'},) type_debug=tuple"
}

TASK [assert that simple_line is string] *********************************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible-bug-repro/lineinfile_json_21065/test.yml:14
fatal: [localhost]: FAILED! => {
    "_ansible_no_log": false, 
    "_ansible_verbose_always": true, 
    "assertion": "simple_line|type_debug != \"tuple\"", 
    "changed": false, 
    "evaluated_to": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "that": "simple_line|type_debug != \"tuple\""
        }, 
        "module_name": "assert"
    }
}
        to retry, use: --limit @/home/adrian/src/ansible-bug-repro/lineinfile_json_21065/test.retry

PLAY RECAP ***************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1   
```

**Comment by bcoca** [comment]
this is to be expected as 'type detection' will see any leading { or [ after templating as a dictionary and list respectively.

one easy way to bypass is adding a leading space

**Comment by photoninger** [comment]
But why does it only detect a dictionary in one of my three lines and not also in the other two? 
I would expect that either all three are detected as dictionaries or none of them.

**Comment by photoninger** [comment]
I think I figured it out: The line gets only converted to a python data type if templating is used AND the resulting string has correct python syntax. 
I managed to change my playbook to "mess up" all three lines, even with quoting my integer variable:
```yaml
- hosts: testserver
  remote_user: root
  vars:
    - ZOOKEEPERPORT: "2222"
    - bla: abc
  tasks:
  
  - name: add line to config
    lineinfile: 
      dest: /root/test_jolokia.cfg
      line: "{\"port\" : 1{{ ZOOKEEPERPORT }}, \"instance\" : \"zookeeper-{{ ZOOKEEPERPORT }}\"}," 
      state: present 
      insertafter: "^instances = "

  - name: add line to config
    lineinfile: 
      dest: /root/test_jolokia.cfg
      line: "{\"port\" : \"{{ bla }}\", \"instance\" : \"zookeeper-{{ bla }}\"}," 
      state: present 
      insertafter: "^instances = "

  - name: add line to config
    lineinfile: 
      dest: /root/test_jolokia.cfg
      line: "{\"port\" : 11111, \"instance\" : \"zookeeper-{{ ZOOKEEPERPORT }}\"}," 
      state: present 
      insertafter: "^instances = "
```
So this behavior is meant by the sentence "templating (variables in playbooks and template lookups) has improved with regard to keeping the original instead of turning everything into a string. If you need the old behavior, quote the value to pass it around as a string." in the https://docs.ansible.com/ansible/porting_guide_2.0.html?
Then it is a feature and not a bug....

**Comment by waxlamp** [comment]
Hi - I am having the same problem with an Ansible playbook, and I'm using the same workaround of a leading space to escape this behavior. I'm worried that the workaround might itself get flagged as a bug and then "fixed" in a future release. Some guidance on the correct way to really defeat the "type detection" would be wonderful.

**Comment by bcoca** [comment]
ending with `|string` or `|to_json` will always bypass 'typing'

**Comment by ansibot** [comment]
cc @ahtik @dhozac
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ToBeReplaced** [comment]
Just a heads up, this also breaks `|to_nice_json` if you pass any arguments to it (ex. `indent=2`), so the only bypass I would rely on is `|string`.

**Comment by madisonb** [comment]
Came across this issue trying to debug a very similar thing.

In our case, we were trying to insert a templated string that was a valid python dictionary into a configuration file, like so:

vars
```
overrides: |-
       {
           "a.hosts": "{% for host in a_hosts %}{{ host }}:{{ a_port|default(5000) }}{% if not loop.last %},{% endif %}{% endfor %}",
           "b.hosts": [{% for host in b_hosts %}"{{ host }}:{{ b_port|default(9000) }}"{% if not loop.last %},{% endif %}{% endfor %}],
           "log.dir": "{{ the_log_dir }}",
           "log.max.bytes": 5 * 1024 * 1024,
           "log.backups": 2,
}
```

templated python config file has within it:
```
OVERRIDES = {{ overrides|default('{}') }}
```

Which resulted in a python dictionary being put into the template file, instead of our nice clean formatted string. Almost as if ansible converted the whole string into a dictionary and just printed `str(dict)`.

As a workaround, you can turn the string into invalid python by appending characters of your choice to the end of it, and then replace the string via a filter.

vars
```
overrides: |-
       {
           "a.hosts": "{% for host in a_hosts %}{{ host }}:{{ a_port|default(5000) }}{% if not loop.last %},{% endif %}{% endfor %}",
           "b.hosts": [{% for host in b_hosts %}"{{ host }}:{{ b_port|default(9000) }}"{% if not loop.last %},{% endif %}{% endfor %}],
           "log.dir": "{{ the_log_dir }}",
           "log.max.bytes": 5 * 1024 * 1024,
           "log.backups": 2,
}badsnippet
```
templated python config file
```
OVERRIDES = {{ overrides|replace('badsnippet','')|default('{}') }}
```
and the result is as you would expect.

The above suggestions of using a filter like `|string` or `|to_json` did not work, and neither did adding a space to the beginning of the string.

**Comment by ansibot** [comment]
cc @samdoran
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by samdoran** [comment]
Duplicate of #22255
